### PR TITLE
s3: a list marker that sorts before the prefix excludes nothing

### LINKED ISSUE #11321: S3: ListObjectsV2 start-after / ListObjects marker that sorts before the prefix returns an empty listing (breaks docker/distribution and zot repository walks)
## Describe the bug

`ListObjectsV2` (and `ListObjects` v1) return an empty result when `start-after` / `marker` sorts lexicographically **before** the `prefix` and is not itself under the prefix. Per the S3 API the marker is a plain key cutoff ("start listing after this key"); every key carrying the prefix already sorts after such a marker, so the response must be the same as with no marker at all.

`normalizePrefixMarker` treats any marker that does not start with the prefix as "something wrong" and the listing comes back with `KeyCount 0`, `IsTruncated false`.

## Why it matters

docker/distribution's S3 storage driver (v3.x, `registry/storage/driver/s3-aws/s3.go`, `doWalk`) always sends `start-after=<rootdirectory>` together with `prefix=<rootdirectory>/<path>/` when it walks storage. On SeaweedFS that walk sees an empty bucket. zot (which uses that driver) therefore lists **no repositories**: `/v2/_catalog` is empty, GC / scrub / retention never see a repo, and on restart zot's storage parse deletes the metadata of every repository "no longer in storage". Pushes and pulls of exact keys keep working, so this is silent until a restart.

## System Setup

- SeaweedFS 4.44 and 4.47 (`chrislusf/seaweedfs:4.47`, `weed server -s3`), reproduced identically on both
- OS: Linux (k3s on Ubuntu 24.04; also a plain `docker run` on the same host)
- Filer store: leveldb2 (default)

## Steps to reproduce

```sh
docker run -d --name sw -p 127.0.0.1:8333:8333 chrislusf/seaweedfs:4.47 server -s3 -ip.bind=0.0.0.0 -volume.max=0
sleep 5; S=http://127.0.0.1:8333
curl -s -X PUT $S/zot
for k in zot/zot/_restore_complete zot/zot/rob/ourea/index.json zot/zot/rob/ourea/oci-layout zot/zot/app/anvil/index.json; do
  curl -s -X PUT --data-binary x "$S/zot/$k"; done

curl -s "$S/zot?list-type=2&prefix=zot/zot/"                       # KeyCount 4  (correct)
curl -s "$S/zot?list-type=2&prefix=zot/zot/&start-after=zot/zot/"  # KeyCount 4  (correct)
curl -s "$S/zot?list-type=2&prefix=zot/zot/&start-after=zot"       # KeyCount 0  (BUG — expected 4)
curl -s "$S/zot?list-type=2&prefix=zot/zot/&start-after=a"         # KeyCount 0  (BUG — expected 4)
curl -s "$S/zot?list-type=2&prefix=zot/zot/&start-after=zot/zot"   # KeyCount 0  (BUG — expected 4)
```

Against AWS S3 (and MinIO) the last three return all four keys.

## Expected behavior

A marker that sorts before the prefix excludes nothing; the listing equals the no-marker listing. A marker that sorts after the prefix's subtree correctly lists nothing (unchanged).

## Additional context

The relevant code is `normalizePrefixMarker` in `weed/s3api/s3api_object_handlers_list.go`:

```go
if !strings.HasPrefix(marker, prefix) {
    // something wrong
    return "", prefix, marker
}
```

PR with a fix and a unit test follows.

### PR DESCRIPTION
**Pull request by robfischer1**
## What problem are we solving?

`ListObjectsV2` `start-after` and `ListObjects` `marker` are a plain key cutoff: return the keys that sort after it. When the marker sorts **before** the prefix and is not under it, every key carrying the prefix already sorts after the marker, so the listing must equal the no-marker listing. `normalizePrefixMarker` treated such a marker as "something wrong" and the listing came back empty (`KeyCount 0`).

docker/distribution's S3 storage driver (v3.x `doWalk`) always sends `start-after=<rootdirectory>` with `prefix=<rootdirectory>/<path>/`. On SeaweedFS that walk saw an empty bucket, and zot — which uses the driver — listed no repositories: `/v2/_catalog` empty, GC/scrub/retention never saw a repo, and on restart its storage parse deleted every repository's metadata as "no longer in storage". Exact-key pushes and pulls kept working, so it was silent until a restart. Fixes #11321.

## How are we solving the problem?

`listFilerEntries` now lists as if no marker were given when `markerSortsBeforePrefix(prefix, marker)` holds. The response still echoes the marker the client sent. A marker that sorts **after** the prefix but is not under it is left alone: it may sit inside a partial-name prefix's match set (`parent` also matches `parentDir/…`), which `normalizePrefixMarker` already handles, and otherwise correctly lists nothing.

## How is the change tested?

- New `Test_markerSortsBeforePrefix` table covers the distribution shape (`zot/zot/` + `zot`), the prefix directory without its slash, unrelated earlier keys, leading slashes, and the negative cases (empty marker, empty prefix, marker under the prefix, marker after the subtree, partial-name prefix).
- `go test ./weed/s3api/` passes.
- End to end against `weed server -s3` built from this branch, versus 4.44 and 4.47 unpatched:

```
prefix=zot/zot/ start-after=zot/zot/   4.47: 5 keys   patched: 5 keys
prefix=zot/zot/ start-after=zot        4.47: 0 keys   patched: 5 keys
prefix=zot/zot/ start-after=a          4.47: 0 keys   patched: 5 keys
prefix=zot/zot/ start-after=zot/zot    4.47: 0 keys   patched: 5 keys
```

A zot registry pointed at the patched build lists its repositories again.

---

<a href="https://app.devin.ai/review/seaweedfs/seaweedfs/pull/11322" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-devin-review-dark.svg?v=3">
    Devin Review
  </picture>
</a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Corrected object listings when the starting marker sorts before the selected prefix.
  * Ensured applicable directory-like keys appear when markers match or precede a prefix.
  * Improved delimiter handling across initial and retried listing requests while preserving marker exclusivity.
  * Preserved client-supplied markers in applicable listings and excluded matching marker keys.

* **Tests**
  * Added coverage for marker and prefix boundary conditions.
  * Added validation for delimiter-adjusted markers and continued pagination.

### CODE REVIEW

File: weed/s3api/s3api_object_handlers_list.go [hunk 1]
```diff
@@ -264,8 +264,15 @@ func (s3a *S3ApiServer) listFilerEntries(ctx context.Context, req listObjectsReq
...
+	// A marker (start-after) that sorts before the prefix excludes nothing under it:
+	// every key carrying the prefix already sorts after the marker. List as if no
+	// marker were given; the response still echoes the marker the client sent.
+	listMarker := originalMarker
+	if markerSortsBeforePrefix(originalPrefix, originalMarker) {
+		listMarker = ""
+	}
 	// convert full path prefix into directory name and prefix for entry name
-	requestDir, prefix, marker := normalizePrefixMarker(originalPrefix, originalMarker)
+	requestDir, prefix, marker := normalizePrefixMarker(originalPrefix, listMarker)
```
[the remark this answers is not in the document]
Comment (chrislusf) [reply weed/s3api/s3api_object_handlers_list.go]: Fixed. The marker was losing its boundary before `listFilerEntries` ever saw it: `adjustMarkerForDelimiter` trimmed `a/b/` to `a/b`, which then sorts before the prefix and got dropped, so the prefix directory key came back. A marker equal to the prefix names no subtree to skip, so it is no longer trimmed — the walk starts inside that directory, its own key stays excluded, and the children still list. Covered by `Test_normalizePrefixMarker` and a listing-level test for both start-after shapes.

File: weed/s3api/s3api_object_handlers_list_marker_before_prefix_test.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// A marker that sorts before the prefix excludes nothing under the prefix, so the
+// listing must be the one with no marker. docker/distribution's S3 driver sends
+// exactly this — prefix "<root>/<path>/" with start-after "<root>" — and an empty
+// answer made a zot registry on SeaweedFS believe it held no repositories.
+func Test_markerSortsBeforePrefix(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (chrislusf) [reply weed/s3api/s3api_object_handlers_list_marker_before_prefix_test.go:13]: Added `TestListWithMarkerBeforePrefix`, which drives the real derivation (`adjustMarkerForDelimiter` + `normalizePrefixMarker` + the cursor flag) into `doListFilerEntries` over the synthetic filer client and asserts the returned keys. `listFilerEntries` itself dials a filer, so the marker echo stays out of reach of a unit test; the derivation it wires is now what the test exercises.

File: weed/s3api/s3api_object_handlers_list.go [hunk 3]
```diff
@@ -1311,18 +1333,19 @@ func compareWithDelimiter(a, b, delimiter string) bool {
 // but still finds any "bop" or later entries. We add a high ASCII character rather than incrementing
 // the last character to avoid skipping potential directory entries.
 // This is essential for correct S3 list operations with delimiters and CommonPrefixes.
-func adjustMarkerForDelimiter(marker, delimiter string) string {
-	if delimiter == "" || !strings.HasSuffix(marker, delimiter) {
+// A marker equal to the prefix names no subtree to skip: it excludes only the prefix's own key.
+func adjustMarkerForDelimiter(marker, prefix, delimiter string) string {
+	if delimiter == "" || marker == prefix || !strings.HasSuffix(marker, delimiter) {
```
[the remark this answers is not in the document]
Comment (chrislusf) [reply weed/s3api/s3api_object_handlers_list.go]: Fixed — `adjustMarkerForDelimiter` now compares the marker and prefix with leading slashes trimmed, the same normalization `normalizePrefixMarker` and `markerSortsBeforePrefix` apply, so `prefix=/logs/` with `marker=logs/` keeps its boundary and the `logs/` key stays excluded. `TestListWithMarkerBeforePrefix` covers the leading-slash marker.

**Comment by chrislusf** [comment]
Addressed the V1 marker echo finding from https://github.com/seaweedfs/seaweedfs/pull/11322#pullrequestreview-5206281287.

`sanitizeV1MarkerEcho` now takes both the marker the client sent and the cutoff the walk used: it echoes the former in `ListBucketResult.Marker` and excludes both from the listing, so `marker=docker/` no longer comes back as `docker` and a `docker/` key the marker names is not re-listed. Verified against a local `weed server -s3`: `?marker=docker%2F&delimiter=/` echoes `<Marker>docker/</Marker>`, and `TestSanitizeV1MarkerEcho_DelimiterAdjustedMarker` pins it.

File: weed/s3api/s3api_object_handlers_list.go [hunk 4]
```diff
@@ -210,18 +211,25 @@ func (s3a *S3ApiServer) ListObjectsV1Handler(w http.ResponseWriter, r *http.Requ
...
 
-	markerCandidates := map[string]struct{}{
-		marker:                          {},
-		strings.TrimPrefix(marker, "/"): {},
-	}
-	if encodingTypeUrl {
-		escapedMarker := urlPathEscape(strings.TrimPrefix(marker, "/"))
-		markerCandidates[escapedMarker] = struct{}{}
+	markerCandidates := map[string]struct{}{}
+	for _, m := range []string{requestMarker, marker} {
```
[the remark this answers is not in the document]
Comment (chrislusf) [reply weed/s3api/s3api_object_handlers_list.go]: Fixed by narrowing the change to the echo. `sanitizeV1MarkerEcho` now filters only the cutoff the walk actually used, as it did before, so no key that spent a slot of the page budget is dropped after the fact and a truncated page cannot be turned into a final empty one. The request marker is used solely for `ListBucketResult.Marker`.

Note the reported case (`prefix=docker/`, `marker=docker/`) does not reach the filter at all: a marker equal to the prefix is no longer trimmed, so the walk starts inside that directory and the `docker/` key is excluded structurally rather than by post-filtering.

**Comment by chrislusf** [comment]
Addressed https://github.com/seaweedfs/seaweedfs/pull/11322#pullrequestreview-5206488092, together with the opposite finding in https://github.com/seaweedfs/seaweedfs/pull/11322#discussion_r4012675281 — post-filtering cannot satisfy both. Filtering the client marker out of a finished page drops a key that already spent a slot of `maxKeys`, and when it was the only key the page held, the no-progress guard has to choose between clearing truncation (later keys lost) and keeping `NextMarker == marker` (the client loops forever).

So the exclusion moved into the walk instead: `listFilerEntries` now receives the marker as the client sent it (`requestMarker`) alongside the walk cutoff, and `excludedMarkerKey` names the one key the trimmed cutoff no longer excludes. That entry is skipped as it streams, so it spends no slot, `nextMarker` still advances past it, and the page fills with real keys. The response echoes `requestMarker` in `ListBucketResult.Marker`, which also fixes the V1 echo for both handlers, so `sanitizeV1MarkerEcho` is back to its original signature and behavior.

Verified against a local `weed server -s3` with `delimiter=|` and keys `docker|`, `docker|x`, `docker|y`, `dockerz`: `?marker=docker%7C&delimiter=%7C` echoes `<Marker>docker|</Marker>`, omits the `docker|` key, and still returns the `docker|` CommonPrefix; with `max-keys=1` the pages walk `dockerz` then `docker|` with no key lost and no repeat. `Test_excludedMarkerKey` covers the decision.
